src/chains/ethereum/ethereum/src/blockchain.ts

```diff
@@ -586,6 +586,71 @@ export default class Blockchain extends Emittery<BlockchainTypedEvents> {
     );
   };
 
+  createPendingBlock = async (previousBlock: Block) => {
+    const options = this.#options;
+    const minerOpts = options.miner;
+    // "instamine" mode means blockTime = 0
+    const instamine = this.#instamine;
+    let timestamp: number = null;
+    if (!instamine) {
+      // we can't always get the pending block's timestamp right because they
+      // could have the miner paused, but if blockTime > 0, blocks are mined at
+      // regular intervals, so we can at least make an educated guess
+      timestamp =
+        previousBlock.header.timestamp.toNumber() + minerOpts.blockTime;
+    }
+    const nextBlock = this.#readyNextBlock(previousBlock, timestamp);
+    // If the executables has inProgress transactions, we were already
+    // in the middle of mining a block when the pending block was requested.
+    // The block that is currently being mined is what we want to return as pending,
+    // so we'll reset the executables to be what it was when mining started,
+    // i.e. all inProgress executables are moved to the front of the pending executables.
```

Comment:
Hmm, I think there's some misunderstandings on how the transaction pool works here, but I could be just misunderstanding you. In researching this PR, I wrote up [some notes](https://hackmd.io/FgAwge5kS26budoVrwn8DA) on hackmd that might be helpful with definitions, but I can elaborate here.

A transaction enters the pool and stays there _until it is included on a block and `"finalized"`_. Up until that point, it stays in the pool.

When a transaction enters the pool, it can be either `executable` or `queued`. Executable transactions are valid transactions that have a nonce of `sender.transactionCount + 1`. Queued transactions are valid transactions that have a nonce that's higher than that.

The executable transactions are further divided in to `pending` and `inProgress`. Pending transactions are waiting to be run. In progress transactions have been run (meaning `vm.runTx` has been called on the transaction), but the block hasn't been finalized yet (maybe cause the miner is still running more transactions to go on that block), so it's kinda in limbo still.

So, for Ganache, a transaction lifecycle is something like:

`pending` => `runTx called` => `removed from pending and added to inProgress` => `block finalized and removed from inProgress`

The transaction is in the transaction pool until the last step.

Reply:
After typing all of that up and rereading your question a few more times, I think I now understand your question. I should listen more and talk less :sweat_smile: 

My interpretation of your question:
If a transaction has been moved from the executable's `pending` transactions to `inProgress` transactions and the `pending` block is requested, will that transaction always be in the `pending` block?

I _think_ the answer here is almost always yes. The transaction will be run again on a new block with identical fields except potentially the `timestamp`. So if one of those transactions that is already mined is a contract call that is somehow dependent on the block's `timestamp`, it could potentially fail on the `pending` block when it didn't originally.

Reply:
Awesome, thanks for the clarification! Definitely helped me.